src/core/cpu_core.cpp

```diff
@@ -627,6 +690,126 @@ void DisassembleAndPrint(u32 addr, u32 instructions_before /* = 0 */, u32 instru
   }
 }
 
+ALWAYS_INLINE_RELEASE static bool ConditionalBreakpointLookAhead(u32 inst_bits)
+{
+  const u32 pc = g_state.regs.pc;
+  Instruction inst = g_state.current_instruction;
+  inst.bits = inst_bits;
+
+  if (System::IsPaused() || pc == s_last_data_breakpoint_check_pc)
+    return false;
+
+  // Skip nops. Makes PGXP-CPU quicker, but also the regular interpreter.
+  if (inst.bits == 0)
+    return false;
+
+  switch (inst.op)
+  {
+    case InstructionOp::lb:
+    case InstructionOp::lbu:
+    {
+      const VirtualMemoryAddress addr = ReadReg(inst.i.rs) + inst.i.imm_sext32();
+      DataBreakpointCheck<MemoryAccessType::Read>(addr);
+    }
+    break;
+    case InstructionOp::lh:
+    case InstructionOp::lhu:
+    {
+      const VirtualMemoryAddress addr = ReadReg(inst.i.rs) + inst.i.imm_sext32();
+      DataBreakpointCheck<MemoryAccessType::Read>(addr);
+      DataBreakpointCheck<MemoryAccessType::Read>(addr + 1);
+    }
+    break;
+    case InstructionOp::lw:
+    {
+      const VirtualMemoryAddress addr = ReadReg(inst.i.rs) + inst.i.imm_sext32();
+      DataBreakpointCheck<MemoryAccessType::Read>(addr);
```

Comment:
This is going to get *very* slow. It would probably be better to align the test address down, and do a single range check in `DataBreakpointCheck`.